### PR TITLE
Fix `.`, `"` error in subfields in map score modifiers

### PR DESCRIPTION
**Pull request by RaynorChavez**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)

Scenario 1: 
- Say we have the document: `{"_id": "1_map", "text_field": "a photo of a cat", "map_score_mods_float": {"a.subsubfield": 0.5}},`
- And we attempt to access the score modifier in a query through: `"add_to_score": [{"field_name": "map_score_mods_float.a.subsubfield", "weight": 2}]`
- Then we will get an error:
```
status_code: 500
ValueError: too many values to unpack (expected 2)
```

Scenario 2:
- with the document `{"_id": "1_map", "text_field": "a photo of a cat", "map_score_mods_float": {"a"subsubfield": 0.5}},`
- we will get an error: ```Failed for special character '"': MarqoWebError: MarqoWebError Error message: {'message': "Marqo encountered an unexpected internal error.\nPlease create an issue on Marqo's GitHub repo (https://github.com/marqo-ai/marqo/issues) if this problem persists.", 'code': None, 'type': None, 'link': ''}
status_code: 500, type: None, code: None, link: ```
- In marqo, the error is: ```marqo.vespa.exceptions.VespaStatusError: 400: {"root":{"id":"toplevel","relevance":1.0,"fields":{"totalCount":0},"errors":[{"code":3,"summary":"Illegal query","message":"Could not set 'ranking.features.query(marqo__add_weights_tensor)' to '{\"map_score_mods_float.a\\\"subsubfield\": 2.0}': Could not parse '{\"map_score_mods_float.a\\\"subsubfield\": 2.0}' as a tensor of type tensor(p{}): At value position 26: Expected a ':' but got 's'"}]}}```

Scenario 3:
- for all subfields unicode characters and `\\` 
- users will be able to index, but will get a silent error when searching because of the discrepancy between the encoding format for these characters between add_docs and the query - thus, search will be possible, but the correct score_modifier will not be matched and applied

* **What is the new behavior (if this is a feature change)?**
- For score modifiers with `.`, expected behavior is that we no longer get an internal error saying too many values to unpack and that the score is correctly applied.
- For score modifiers with `"`, expected behavior is that we no longer have the 500 internal error when searching and ~~that the score modifier is applied to the score~~ As discussed, we will let this silent error like other unicode characters until we hear back from the vespa team. Currently, we will encode " in unicode. I have confirmed that this will not match with documents explicitly containing the `\u0022` encoding.
- For all other unicode characters,  we will let these silent fail for now.

The main changes to enable this were:
- For `.` we simply split the query along the first `.` character, resulting in only 2 partitions, instead of in all `.`, which may result in multiple partitions if the query score modifier has more that one `.`
- For `"`, Vespa seems to have a bug wherein it attempts to escape `"` in the queries by transforming it to `\\\"`, leading to an invalid query. ~~To solve this without modifying vespa code, we switch to url encoding (where `"` is stored as `%22`) in both documents and queries. Note that this only applies to the subfields of score modifier fields in both unstructured and structured indexes~~ . Use `re.sub(r'(")', lambda m: r'\u0022', s)` .
- 
* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
 - Need someone to check

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)

* **Related Python client changes** (link commit/PR here)

* **Related documentation changes** (link commit/PR here)

* **Other information**:

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/core/utils/special_characters_encoding.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import re
+
+def custom_encode(s: str) -> str:
```
Comment (wanliAlex) [inline src/marqo/core/utils/special_characters_encoding.py]: Can you provide some doc strings for these functions?

File: tests/core/structured_vespa_index/test_structured_vespa_index.py
Comment (wanliAlex) [inline tests/core/structured_vespa_index/test_structured_vespa_index.py:1]: More tests about `"` are also needed?

**Review by wanliAlex — CHANGES_REQUESTED** [review]

File: src/marqo/core/utils/special_characters_encoding.py (hunk 1 above)
Comment (RaynorChavez) [reply src/marqo/core/utils/special_characters_encoding.py]: Added!

File: tests/core/structured_vespa_index/test_structured_vespa_index.py
Comment (RaynorChavez) [reply tests/core/structured_vespa_index/test_structured_vespa_index.py:1]: Yes, I had them before, but removed due to transient local integ test failing issue. I thought my tests were the problem.

I've re-added these now

**Review by wanliAlex — DISMISSED** [review]
LGTM

File: tests/tensor_search/integ_tests/test_search_unstructured.py [hunk 2]
```diff
@@ -1441,3 +1442,116 @@ def run():
                       vespa_query_kwargs["yql"])
         self.assertIn("embedding_query", vespa_query_kwargs["query_features"])
         self.assertNotIn("marqo__query_embedding", vespa_query_kwargs["query_features"])
+
+    def test_special_characters_in_map_score_modifiers(self):
+         special_characters = [
```
Comment (papa99do) [inline tests/tensor_search/integ_tests/test_search_unstructured.py]: nit. the special chars are duplicated in three tests, can we extract it?

File: tests/tensor_search/integ_tests/test_search_structured.py [hunk 3]
```diff
@@ -1184,3 +1186,115 @@ def test_tensor_search_with_custom_vector_query(self):
...
+                     index_name=self.default_text_index,
+                     text="",
+                     score_modifiers=score_modifiers,
+                 )
+
+                 expected_score = 1.4165449318484857
+                 actual_score = res['hits'][0]['_score']
+
+                 if abs(actual_score - expected_score) > 0.01:
+                     #print(f"Soft fail for special character '{special_character}': Score mismatch")
```
Comment (papa99do) [inline tests/tensor_search/integ_tests/test_search_structured.py]: can we remove the commented out code from tests?

Comment (RaynorChavez) [reply tests/tensor_search/integ_tests/test_search_structured.py]: I wanted to leave them commented out for future debugging. What do you think?

File: tests/tensor_search/integ_tests/test_search_unstructured.py (hunk 2 above)
Comment (RaynorChavez) [reply tests/tensor_search/integ_tests/test_search_unstructured.py]: I'll put it in a common file, thanks

File: src/marqo/core/utils/special_characters_encoding.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
```
Comment (farshidz) [inline src/marqo/core/utils/special_characters_encoding.py]: Why not have these in the base vespa index class?

File: src/marqo/core/vespa_index/vespa_index.py [hunk 5]
```diff
@@ -129,20 +130,24 @@ def _get_score_modifiers(self, marqo_query: MarqoQuery) -> Optional[Dict[str, Di
         if marqo_query.score_modifiers:
             mult_tensor, add_tensor = self._convert_score_modifiers_to_tensors(marqo_query.score_modifiers)
```
Comment (farshidz) [inline src/marqo/core/vespa_index/vespa_index.py]: You can encode in this function to avoid an extra pass over the fields

File: tests/core/structured_vespa_index/test_structured_vespa_index.py [hunk 6]
```diff
@@ -251,3 +252,39 @@ def test_to_vespa_document_fieldNotInIndex_fails(self):
...
+    def test_score_modifier_with_dot_in_subfield(self):
+        """
+        Test case where the score modifier field contains a period ('.') in the subfield name.
+        Ensures that only the first occurrence of '.' is split, and the rest is retained as part of the subfield.
+        """
+        marqo_doc = {
+            '_id': 'my_id',
+            'title': 'my title',
+            'price': 100.0,
+            "map_float_field": {"sub.field": 1.23},
```
Comment (farshidz) [inline tests/core/structured_vespa_index/test_structured_vespa_index.py:264]: Test one dot as edge case (as you have) but also test more than one as the average case

File: tests/core/structured_vespa_index/test_structured_vespa_index.py [hunk 7]
```diff
@@ -251,3 +252,39 @@ def test_to_vespa_document_fieldNotInIndex_fails(self):
...
+                'title': {
+                    constants.MARQO_DOC_CHUNKS: ['my', 'title'],
+                    constants.MARQO_DOC_EMBEDDINGS: [[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]
+                }
+            }
+        }
+
+        actual_vespa_doc = self.vespa_index.to_vespa_document(marqo_doc)
+        expected_vespa_doc = {
+            'id': 'my_id',
```
Comment (farshidz) [inline tests/core/structured_vespa_index/test_structured_vespa_index.py]: You could simplify your test here and just check the ID of the doc returned. Focused tests are more maintainable

File: src/marqo/core/utils/special_characters_encoding.py (hunk 4 above)
Comment (RaynorChavez) [reply src/marqo/core/utils/special_characters_encoding.py]: Previously also used them to encode add docs to url encoding, so it made sense to separate. But now that we only use it for queries, you're right it's better to just put it in the base vespa index class.

Moved!

File: src/marqo/core/vespa_index/vespa_index.py (hunk 5 above)
Comment (RaynorChavez) [reply src/marqo/core/vespa_index/vespa_index.py]: Thanks, done

File: tests/core/structured_vespa_index/test_structured_vespa_index.py (hunk 6 above)
Comment (RaynorChavez) [reply tests/core/structured_vespa_index/test_structured_vespa_index.py:264]: Added a new test

File: tests/core/structured_vespa_index/test_structured_vespa_index.py (hunk 7 above)
Comment (RaynorChavez) [reply tests/core/structured_vespa_index/test_structured_vespa_index.py]: Thanks, simplified

File: tests/tensor_search/integ_tests/test_search_unstructured.py (hunk 2 above)
Comment (RaynorChavez) [reply tests/tensor_search/integ_tests/test_search_unstructured.py]: Done

File: tests/tensor_search/integ_tests/test_search_structured.py (hunk 3 above)
Comment (RaynorChavez) [reply tests/tensor_search/integ_tests/test_search_structured.py]: Removed!

**Review by farshidz — APPROVED** [review]
